### DISCUSSION TITLE
Heat source location set problem

### DISCUSSION THREAD

**Discussion by dpokras** [pysal/spopt#236] [Q&A]
Hello!

Thanks for the library, I think it has potential in carrying my project forward. I'm a researcher at ETH Zurich working on the PARTHFNDR project. My current research revolves around modeling the use of waste heat resources coupled with renewable energy resources in rolling-out Direct Air CO2 capture (DAC) plants in Europe. 

For one scenario, I think the LSCP model would be quite useful in helping me calculate the locations of the DAC sites in-between the heat sources. I would like to know where to place the minimal number of DAC sites in-between the heat sources so that all the heat demand is covered. However, the heat sources themselves have distance constraints, meaning the heat cannot be transported further than the maximum feasible range. Each heat source has a different range so its not as simple as adding one maximal_coverage_constraint. I've attached a picture of what this looks like - these are the red points with the green buffers (please download the image and zoom into the a cluster area to understand better). Is there a way to modify the current model to include these constraints, such as inputting a list instead of a scalar into the maximal_coverage_constraint?
![resources](https://user-images.githubusercontent.com/105276214/170220475-38a698b9-4c35-42ee-bb3e-6828eb03dc83.png)

My current approach was to create a meshgrid bounded within the buffers (maximal heat source ranges) where all the possible DAC sites may be located. I've also attached what this looks like.
<img width="563" alt="meshgrid_heat_sources" src="https://user-images.githubusercontent.com/105276214/170220656-393f71e7-1763-43e3-99ba-53f7ba8b0494.png">

**Comment by gegen07** [comment]
Hmm... Currently spopt doesn't solve this problem, unfortunately. Firstly you cannot add a maximal_constraint in spopt LSCP model. Besides that, you cannot input a list of ranges in this constraint even in MCLP. 

@jGaboardi @qszhao this type of problem is so specific or it's ok to change/add parameters in spopt coverage models to support these constraints?

**Reply by qszhao** [reply]
It looks like with the current meshgrid solution, it can be just converted to a traditional LSCP problem?

**Reply by dpokras** [reply]
@qszhao this was my idea, however, some heat sources wouldn't be able to reach all potential DAC sites. I thought of fixing this by creating a network in which the "streets" are all allowable connections in the system. This is a picture of one such network: 
![galicia nodes](https://user-images.githubusercontent.com/105276214/170693031-133d14ed-b558-4a19-be17-53db3281d6ad.png)

Is it possible to import such a network into the LSCP optimiser? (with an associated cost matrix)

**Reply by gegen07** [reply]
Yes, @dpokras! Take a look at [here](https://pysal.org/spopt/notebooks/lscp.html#Calculating-the-cost-matrix). It explains how to input a precomputed cost matrix into LSCP.

**Reply by dpokras** [reply]
I managed to run a LSCP optimisation with the network used above. I had to modify the plotting function mainly because there isnt enough colors to label every point. Is there a way to increase the amount of colors used?
![Figure 2022-05-31 205019](https://user-images.githubusercontent.com/105276214/171283828-eabfe968-b3a8-40e7-a859-ee4da78d5845.png)


Additionally, I need to extract the demand that is covered by each located point, in order to determine the overall cost for my network. Could I extract the LSCP points as a geopandas dataframe?

**Reply by jGaboardi** [reply]
@dpokras As for your first question, you can simply add [more colors](https://matplotlib.org/stable/gallery/color/named_colors.html) to `dv_colors_arr` in cell 18 of the link @gegen07 posted above.

**Reply by dpokras** [reply]
Regarding the second question, I understand that I can extract the geometry points for the facility locations as well as the points for the covered and uncovered demand locations. However, I need to be able to extract the total demand covered by each facility location and from which demand point (and how much from each demand point).

Would you be able to guide me in extracting this information?

**Reply by gegen07** [reply]
@dpokras you can use [fac2cli](https://github.com/pysal/spopt/blob/main/spopt/locate/coverage.py#L248) array and [cli2fac](https://github.com/pysal/spopt/blob/main/spopt/locate/base.py#L42) array to get these statistics.
